File: 07. Datos abiertos (Open Data)/MDS_2021_2022_Datos_abiertos_(Open_data)/code/rvest_eurostoxx/scraper_eurostoxx.R
library(rvest)
library(lubridate)
library(tidyverse)
library(httr)

process_timestamp <- function (dates) {
  dates <- str_replace_all(
    dates, 
    c('^([0-9]{1,2}/[0-9]{1,2})$' = paste(paste('\\1', 
                                                as.character(year(Sys.Date())), 
                                                sep = '/'), '23:59'), 
      '^([0-9]{1,2}:[0-9]{1,2})$' = paste(format(Sys.Date(), '%d/%m/%Y'), '\\1')
    )
  )
  dates <- dmy_hm(dates)
  dates <- replace_na(dates, Sys.time())
  
  return(dates)
}

scrape_eurostoxx_data <- function(verbose) {
  url_expansion <- 'http://www.expansion.com/mercados/cotizaciones/indices/eurostoxx_I.5E.html'
  eurostoxx_table_columns <- c('stock', 'last', 'var_perc', 'var', 'ytd', 'max', 
                               'min', 'volume', 'capitalization', 'timestamp')

  expansion_request <- GET(url_expansion)
  
  if (expansion_request$status_code != 200) {
    eurostoxx_table <- data.frame(character(), numeric(), numeric(), numeric(), 
                                  numeric(), numeric(), numeric(), numeric(), 
                                  numeric(), POSIXct())
    
    colnames(eurostoxx_table) <- eurostoxx_table_columns
    return(eurostoxx_table)
  }
  
  expansion_web <- read_html(content(expansion_request, as = 'text'), 
                             encoding = 'ISO-8859-15')

  expansion_tables <- html_nodes(expansion_web, 'table#listado_valores')
  
  eurostoxx_table <- html_table(expansion_tables[[1]], fill = TRUE, dec = ',')

  if (verbose) {
    print(head(eurostoxx_table))
  }

  eurostoxx_table <- eurostoxx_table[, 1:ncol(eurostoxx_table) - 1]
  
  colnames(eurostoxx_table) <- eurostoxx_table_columns
  
  eurostoxx_table <- eurostoxx_table %>% 
    as_tibble()
  
  if (verbose) {
    print(sapply(eurostoxx_table, class))
  }
  
  eurostoxx_table <- eurostoxx_table %>%
    mutate(
      across(
        c(last, max, min, volume, capitalization), 
        ~parse_number(.x, locale = locale(decimal_mark = ',', grouping_mark = '.'))
      )
    )
  
  eurostoxx_table <- eurostoxx_table %>%
    mutate(across(c(var_perc, ytd), function(x) x / 100))
  
  eurostoxx_table <- eurostoxx_table %>%
    mutate(timestamp = process_timestamp(timestamp))
  
  if (verbose) {
    print(sapply(eurostoxx_table, class))
  }
  return(eurostoxx_table)
}

eurostoxx_data <- scrape_eurostoxx_data(T)

write_csv(eurostoxx_data, '.\\rvest_eurostoxx\\eurostoxx.csv' )

File: 07. Datos abiertos (Open Data)/MDS_2021_2022_Datos_abiertos_(Open_data)/code/beautifulsoup_eurostoxx/scraper_eurostoxx.py
import re
from datetime import datetime

import numpy as np
import pandas as pd
import requests
from bs4 import BeautifulSoup


def process_timestamp(element_str):
    timestamp_format = '%d/%m/%Y %H:%M'
    if (pd.isnull(element_str)):
        element = np.nan
    elif (re.search('^[0-9]{1,2}/[0-9]{1,2}$', element_str) != None):
        element_str = element_str + '/' + datetime.now().strftime('%Y') + ' 23:59'
        element = datetime.strptime(element_str, timestamp_format)
    elif (re.search('^[0-9]{1,2}:[0-9]{1,2}$', element_str) != None):
        element_str = datetime.now().strftime('%d/%m/%Y') + ' ' + element_str
        element = datetime.strptime(element_str, timestamp_format)
    else:
        element = datetime.now()

    return element


def scrape_eurostoxx_data(verbose):
    url_expansion = 'http://www.expansion.com/mercados/cotizaciones/indices/eurostoxx_I.5E.html'
    eurostoxx_table_columns = ['stock', 'last', 'var_perc', 'var', 'ytd', 'max', 'min', 'volume',
                               'capitalization', 'timestamp']

    expansion_request = requests.get(url_expansion)

    if expansion_request.status_code != 200:
        return pd.DataFrame([], columns=eurostoxx_table_columns)

    expansion_web = BeautifulSoup(expansion_request.text, 'html.parser')

    if (verbose):
        print(expansion_web.prettify())

    table = expansion_web.find('table', {'id': 'listado_valores'})

    if (verbose):
        print(table.prettify())

    eurostoxx_table = pd.read_html(str(table), 
                                   header = 0, 
                                   encoding = 'utf-8', 
                                   decimal = ',',
                                   thousands = '.')[0]

    if (verbose):
        print(eurostoxx_table.head())

    # Last column is empty so remove it
    eurostoxx_table = eurostoxx_table.iloc[:, :-1]

    # Set new column names
    eurostoxx_table.columns = eurostoxx_table_columns

    if (verbose):
        print(eurostoxx_table.dtypes)

    eurostoxx_table[['var_perc', 'ytd']] = eurostoxx_table[['var_perc', 'ytd']].div(100)

    eurostoxx_table['timestamp'] = eurostoxx_table['timestamp'].apply(process_timestamp)

    if (verbose):
        print(eurostoxx_table.dtypes)

    return(eurostoxx_table)


eurostoxx_data = scrape_eurostoxx_data(True)

eurostoxx_data.to_csv('.\\beautifulsoup_eurostoxx\\eurostoxx.csv', index = False)
eurostoxx_data.to_json('.\\beautifulsoup_eurostoxx\\eurostoxx.json', orient='records', indent = 4, force_ascii = False)
